subprojects/launcher/src/main/java/org/gradle/launcher/daemon/configuration/BuildProcess.java

```diff
@@ -43,9 +43,19 @@ public boolean configureForBuild(DaemonParameters requiredBuildParameters) {
 
         final JvmOptions jvmOptions = new JvmOptions(new IdentityFileResolver());
         jvmOptions.systemProperties(getJvmOptions().getImmutableSystemProperties());
+        if (requiredBuildParameters.hasUserConfiguredImmutableJvmArgs()) {
+            jvmOptions.setJvmArgs(getJvmOptions().getJvmArgs());
+            jvmOptions.setMinHeapSize(getJvmOptions().getMinHeapSize());
+            jvmOptions.setMaxHeapSize(getJvmOptions().getMaxHeapSize());
+        }
+
         List<String> currentImmutables = jvmOptions.getAllImmutableJvmArgs();
         List<String> requiredImmutables = requiredBuildParameters.getEffectiveSingleUseJvmArgs();
-        requiredImmutables.removeAll(DaemonParameters.DEFAULT_JVM_ARGS);
+
+        if (!requiredBuildParameters.hasUserConfiguredImmutableJvmArgs()) {
```

Comment:
Which tests break if we remove both these if-conditions and just compare `getJvmOptions().getAllImmutableJvmArgs()` vs. `requiredBuildParameters.getEffectiveSingleUseJvmArgs() - defaults`?